### ISSUE TITLE
ReinforceSampler update Gradient issue

### ISSUE THREAD

**Issue by wenhuchen** [facebookresearch/MIXER#6]
The training runs into problem after a few epochs, can you help me with that?

Epoch: 1. Training time: 731.75s. WordsXE/s: 3586.04, WordsRF/s: 53.08
Training: Ent: 6.79167 || Ppl: 110.78875 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 6.00815 || Ppl: 64.36271
Epoch: 2. Training time: 711.39s. WordsXE/s: 3688.66, WordsRF/s: 54.60
Training: Ent: 5.42869 || Ppl: 43.07238 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 5.50559 || Ppl: 45.43051
Epoch: 3. Training time: 712.21s. WordsXE/s: 3684.43, WordsRF/s: 54.54
Training: Ent: 4.97243 || Ppl: 31.39422 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 5.15709 || Ppl: 35.68114
Epoch: 4. Training time: 710.02s. WordsXE/s: 3695.77, WordsRF/s: 54.71
Training: Ent: 4.67853 || Ppl: 25.60814 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.93649 || Ppl: 30.62186
Epoch: 5. Training time: 713.61s. WordsXE/s: 3677.21, WordsRF/s: 54.43
Training: Ent: 4.46283 || Ppl: 22.05195 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.73206 || Ppl: 26.57623
Epoch: 6. Training time: 711.31s. WordsXE/s: 3689.08, WordsRF/s: 54.61
Training: Ent: 4.29383 || Ppl: 19.61420 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.62939 || Ppl: 24.75053
Epoch: 7. Training time: 707.76s. WordsXE/s: 3707.59, WordsRF/s: 54.88
Training: Ent: 4.15686 || Ppl: 17.83768 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.62608 || Ppl: 24.69392
Epoch: 8. Training time: 716.06s. WordsXE/s: 3664.60, WordsRF/s: 54.24
Training: Ent: 4.04386 || Ppl: 16.49387 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.61798 || Ppl: 24.55557
Epoch: 9. Training time: 712.12s. WordsXE/s: 3684.88, WordsRF/s: 54.54
Training: Ent: 3.95212 || Ppl: 15.47767 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.45345 || Ppl: 21.90900
Epoch: 10. Training time: 722.68s. WordsXE/s: 3631.05, WordsRF/s: 53.75
Training: Ent: 3.86949 || Ppl: 14.61611 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.43678 || Ppl: 21.65734
Epoch: 11. Training time: 714.16s. WordsXE/s: 3674.35, WordsRF/s: 54.39
Training: Ent: 3.80143 || Ppl: 13.94266 || Avg. reward/token: 0.000 || Cum. reward error: 0.000 || Nsamples Xent: 2624082 || Nsamples Rf: 38842
Validation: Ent: 4.43160 || Ppl: 21.57964/data/users/v-wenhch/torch/install/bin/luajit: bad argument #2 to '?' (out of range)
stack traceback:
        [C]: at 0x7facb696f2f0
        [C]: in function '__newindex'
        /data/users/v-wenhch/MIXER/ReinforceSampler.lua:54: in function 'updateGradInput'
        ...v-wenhch/torch/install/share/lua/5.1/nngraph/gmodule.lua:420: in function 'neteval'
        ...v-wenhch/torch/install/share/lua/5.1/nngraph/gmodule.lua:454: in function 'updateGradInput'
        ...users/v-wenhch/torch/install/share/lua/5.1/nn/Module.lua:31: in function 'backward'
        /data/users/v-wenhch/MIXER/Mixer.lua:289: in function 'train_one_batch'
        /data/users/v-wenhch/MIXER/Trainer.lua:67: in function 'train'
        /data/users/v-wenhch/MIXER/Trainer.lua:134: in function 'run'
        main.lua:176: in main chunk
        [C]: in function 'dofile'
        ...nhch/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
        [C]: at 0x00405d50

**Comment by ranzato** [comment]
Hi, sorry for that!
Are you running the main with the default options? did you change anything in the code?
Please, let me know.
Thank you.

**Comment by wenhuchen** [comment]
Hi, I'm running the main.lua with default setting. I made a very slight changes to the indents, never touched the code itself. I found that the problem is caused by "self.output = torch.multinomial(self.prob, 1 )" and "self.gradInput[ss][self.output[ss][1]] = gradOutput[ss][1]" in ReinfroceSampler.lua. It sometimes outputs very large numbers like 1114098622464, which causes the out of range problem.

Btw, I found that the perplexity on validation set keeps increasing when switching from log-likelihood training to reinforcement training. Is it the normal case?

**Comment by ranzato** [comment]
I think the issue is now fixed. Could you please update to the latest version of Torch and the latest version of this package?

The issue was due to torch.multinomial which had a bug, very very rarely returning an output outside the range. Unfortunately, the new version of torch has changed the type returned by torch.multinomial, so I had to add some casting in ReinforceSampler.

I am rerunning the code and so far I do not see any issue. I will close the task now, but please do let me know if you still have issues (after updating both torch and MIXER).

Thanks for reporting this!
